### ISSUE TITLE
Issues with x32/x64 and module loading

### ISSUE THREAD

**Issue by rfennell** [rfennell/vNextBuild#204]
There is an issue with which pester versions is loaded by the task, it appears to be related to the x32/x64 reload logic, but also highlighed by the changes between Pester 3.4.0 and 4.8.0.

As part of this engineering investigation  (currently a private build) changes have been made to the logic on how Pester is discovered and loaded 
-	A new override flag forces the use of the specified pester version shipped in the task 
-	Next it tests if no module path specific, and no Pester on the agent to use the version in the tasks
-	Next it tries to uses a specified module path
-	Finally tries any default installed on the agent

There are more unit tests for these load models, some of these test do require Pester 4.0.8

Given these changes he issue appears to be 
-	Open X64 powershell and load Pester 4.0.8 all tests pass and can be run repeatedly run
-	If you load in x32 Powershell load 4.0.8 a couple of tests fails (seems for these tests it has fallen back to 3.4.0 of pester), but on second test run most fail, seems to fall back to 3.4.0 of pester completely
-	If you run in x64 and use the force x32 flag seems to always use Pester 3.4.0.

I suspect that we need to force 4.0.8 for the unit tests

Related to #202

**Comment by rfennell** [comment]
I think a sensible issue might be to swap the checkbox for x86 to a drop down 

- Use default arch
- Use x86
- Use AMD64

With some guard logic like

```
write-warning "Processor Architecture $env:Processor_Architecture"
if (($env:PROCESSOR_ARCHITEW6432 -eq "x86") -and ($selectionvar -eq "AMD64")) {
    write-warning "Reloading as 64-bit process"
    &"$env:windir\sysnative\windowspowershell\v1.0\powershell.exe" -noprofile -executionpolicy bypass -file $myinvocation.Mycommand.path $args
    exit $lastexitcode
} 
if (($env:PROCESSOR_ARCHITEW6432 -eq "AMD86")-and ($selectionvar -eq "x86")) {
    write-warning "Reloading as 64-bit process"
    &"$env:windir\sysnative\windowspowershell\v1.0\powershell.exe" -noprofile -executionpolicy bypass -file $myinvocation.Mycommand.path $args
    exit $lastexitcode
} 
```

**Comment by rfennell** [comment]
I have done some more testing.

- If I am in a new 64Bit Powershell windows with only the default modules loaded, all the tests work on the first and every run. They automatically load Pester 4.0.8 and only this version
- If i am in a new 32bit Powershell windows with only the default modules loaded, many the tests fail on the first and every run. Note that invoke-pester automatically loads Pester 3.0.4. If you manually load 4.0.8 of pester most the tests pass, but it seems to unload 4.0.8 so on a second run loads of tests fail

I think the issue is we need to make sure on 32bit that the correct version of Pester is used

**Comment by rfennell** [comment]
If you pull the latest pesterengineering branch  https://github.com/rfennell/vNextBuild/commit/fc5714ac41b775977abf4af23978d532e054d6ef 

- I have sorted the unit tests so the mocking works irrespective of the folder you are in for the module loading related tests
- But there is still the issue that the tests only pass in a 64bit Powershell. If they are run in a 32bit windows, or you force 32bit (with the advanced parameter) inside a build,  or run on a hosted build agent (which ends up as 32bit it seems) it fails as it picks up Pester 3.4.0 as opposed to 4.0.8

My thoughts are we need to do one or more of the following
1. Work out how to make sure invoke-pester runs 4.0.8 (or some specified version) via a parameter – best option if possible
1. In the VSTS task make it a radio button for 32/64 and force the path of the powershell instance – I think this is next thing to try
1. Reload pester install invoke-pester to force a version (not even sure if this can be done)

**Comment by rfennell** [comment]
@ChrisLGardner I think changing the **use32Bit** task.json to a radio button is a good idea. Something like the code sample below. 

My concern is that this is breaking change over the current boolean **use32bit** flag - but I think it might be the answer

```
{
            "name": "platform",
            "aliases": [ "javaHomeOption" ],
            "type": "radio",
            "label": "PowerShell instance architecture",
            "required": true,
            "groupName": "advanced",
            "defaultValue": "default",
            "helpMarkDown": "PowerShell instance architecture e.g. 32/64bit or system default to run tests as.",
            "options": {
                "Default": "Default",
                "x86": "x86",
                "x64": "x64"
            }
        }
```

**Comment by ChrisLGardner** [comment]
Looking at the original task it looks like the logic was always defaulting to the installed version of Pester (usually 3.4.0) unless you specify either a ModuleFolder or Pester isn't installed. I'm not sure how this was then picking up the correct version of the module on x64 unless a newer version than 3.4.0 was installed on any machine this was tested on.

Testing on my machine with the -Run32Bit flag was always loading the latest version of the module (4.1.0 for me), regardless of if I'm running in a 32 bit PS process already or not.

I'm trying the new version of the task on a x86 machine and trying to figure out which version it's loading, I'm hitting some issues with the validation on the ModuleFolder property right now, it looks like it's not correctly handing the [String]::IsNullOrWhiteSpace check so I'm delving into it with Trace-Command to figure out why and then I'll run the task a bunch more times to see what I can find. Running the old task on this machine throws a lot of errors because it's loading the installed version (3.4.0) rather than the task version for the reasons mentioned above.

**Comment by ChrisLGardner** [comment]
So after some testing it looks like the problem I was having was due to calling the pester.ps1 script from the level above it with the following command line:

    .\task\pester.ps1 -ScriptFolder .\test -ResultsFile .\tests.xml -PesterVersion 4.0.8 -ForceUseOfPesterInTasks $True

This was causing some weirdness with the validation on ModuleFolder, making it think I'd passed it `4.0.8` and then failing because there wasn't a folder called 4.0.8 at the level I was running from. Running it from the same directory as the pester.ps1 script with the same parameters was allowing it to run but hitting the same problem you're seeing with the installed version being loaded even if I specify a module version. I've no idea why it was thinking I was passing that value into that parameter rather than the one I'd specified, I might play around with the order the parameters are in to see if the validation is somehow causing it to pick up the value passed into the one before it in the script declaration.

When specifying a version to use and using the -Force flag it seems to run the tests correctly until the last file, where it reverts to Pester 3.4.0 and fails because of the use of new assertions. I'm going to try to dig deeper into why it's swapping to the other version of Pester for that one file and if it always occurs for just that file.

**Comment by ChrisLGardner** [comment]
Published an updated private build which seems to be passing all the tests on a hosted agent. There might be more tests that are needed to double check everything is working out but the latest build should be good to go if everything works out.

**Comment by rfennell** [comment]
Released in 6.1.36
